050-Kafka-Node-Pools.md

```diff
@@ -0,0 +1,806 @@
+
+### Backwards compatibility
+
+The node pools are not backwards compatible, but provide a simple migration path.
+When the feature gate is enabled, users have to use the `KafkaNodePool` resources.
+But they can easily convert the existing `Kafka` custom resource into a matching `Kafka` and `KafkaNodePool` resources.
+For example the following `Kafka` CR which can be used with the feature gate disabled:
+
+```yaml=
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    version: 3.3.1
+    replicas: 3
+    listeners:
+      - name: plain
+        port: 9092
+        type: internal
+        tls: false
+      - name: tls
+        port: 9093
+        type: internal
+        tls: true
+    config:
+      offsets.topic.replication.factor: 3
+      transaction.state.log.replication.factor: 3
+      transaction.state.log.min.isr: 2
+      default.replication.factor: 3
+      min.insync.replicas: 2
+      inter.broker.protocol.version: "3.3"
+    resources:
+      requests:
+        memory: 64Gi
+        cpu: 8
+      limits:
+        memory: 64Gi
+        cpu: 12
+    storage:
+      type: jbod
+      volumes:
+        - id: 0
+          type: persistent-claim
+          size: 1000Gi
+          class: fast-storage
+  # ...
+```
+
+Is equivalent to this combination of `Kafka` and `KafkaNodePool` resources:
+
+```yaml=
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    nodePoolSelector:
+      matchLabels:
+        cluster: my-cluster
+    version: 3.3.1
+    listeners:
+      - name: plain
+        port: 9092
+        type: internal
+        tls: false
+      - name: tls
+        port: 9093
+        type: internal
+        tls: true
+    config:
+      offsets.topic.replication.factor: 3
+      transaction.state.log.replication.factor: 3
+      transaction.state.log.min.isr: 2
+      default.replication.factor: 3
+      min.insync.replicas: 2
+      inter.broker.protocol.version: "3.3"
+  # ...
+---
+
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: kafka
+  labels:
+    cluster: my-cluster
+spec:
+  replicas: 3
+  resources:
+    requests:
+      memory: 64Gi
+      cpu: 8
+    limits:
+      memory: 64Gi
+      cpu: 12
+  storage:
+    type: jbod
+    volumes:
+    - id: 0
+      type: persistent-claim
+      size: 1000Gi
+      class: fast-storage
+```
+
+Because the `KafkaNodePool` is named `kafka`, all the resources created by this configuration will be exactly the same and have the same names as today.
+
+## Open Points
+
+### Migration
+
+The backwards compatibility / migration story seems weak.
+The code already supports handling the original `Kafka` CR through the _virtual node pool_.
+Maybe we can use this and just find a way how to indicate if the user wants to use the `KafkaNodePool` resources or use the `Kafka` with the virtual node pool.
```

Comment:
Without using the feature gate, can we think about something like this? 

A new Strimzi release starts to use the virtual node pool by default so that the code is handling node pools but still driven through only the `Kafka` CR.
If the user wants to explicitly move to use node pools, he creates one or more `KafkaNodePool` custom resources and on next reconcile, the operator will detect their existence and starts using them instead of the virtual one.